orderbook/src/bad_token.rs

```diff
@@ -0,0 +1,465 @@
+use crate::trace_many;
+use anyhow::{anyhow, bail, ensure, Result};
+use contracts::ERC20;
+use ethcontract::{
+    batch::CallBatch,
+    dyns::DynTransport,
+    errors::{ExecutionError, MethodError},
+    transaction::TransactionBuilder,
+};
+use model::TokenPair;
+use primitive_types::{H160, U256};
+use shared::{amm_pair_provider::AmmPairProvider, Web3};
+use std::collections::HashSet;
+use web3::types::{BlockTrace, CallRequest, Res};
+
+#[derive(Debug, Clone)]
+pub enum TokenInfo {
+    Good { gas_per_transfer: U256 },
+    Bad { reason: String },
+    NoPool,
+}
+
+impl TokenInfo {
+    pub fn is_good(&self) -> bool {
+        matches!(self, Self::Good { .. })
+    }
+}
+
+/// Detects whether a token is "bad" (works in unexpected ways that are problematic for solving) by
+/// simulating several transfers of a token. To find an initial address to transfer from we use
+/// the amm pair providers.
+/// Tokens are bad if:
+/// - we cannot find an amm pool of the token to one of the base tokens
+/// - transfer into the settlement contract or back out fails
+/// - a transfer loses total balance
+pub struct BadTokenDetector {
+    pub web3: Web3,
+    pub pools: Vec<Box<dyn AmmPairProvider>>,
+    pub base_tokens: HashSet<H160>,
+    pub settlement_contract: H160,
+}
+
+impl BadTokenDetector {
+    pub async fn detect(&self, token: H160) -> Result<TokenInfo> {
+        let instance = ERC20::at(&self.web3, token);
+        let decimals = match instance.decimals().call().await {
+            Ok(decimals) => decimals,
+            Err(err) => {
+                return match EthcontractError::classify(&err) {
+                    EthcontractError::Node => Err(err.into()),
+                    EthcontractError::Contract => Ok(TokenInfo::Bad {
+                        reason: "failed to get decimals".to_string(),
+                    }),
+                }
+            }
+        };
+        let amount = match U256::from(10).checked_pow(decimals.into()) {
+            Some(amount) => {
+                tracing::debug!("testing token {:?} with amount {}", token, amount);
+                amount
+            }
+            None => {
+                return Ok(TokenInfo::Bad {
+                    reason: "decimals too large".to_string(),
+                });
+            }
+        };
+
+        let take_from = match self.find_address_owning_token(token).await? {
+            Some((address, balance)) if balance > amount => {
+                tracing::debug!("testing token {:?} with pool {:?}", token, address);
+                address
+            }
+            _ => {
+                return Ok(TokenInfo::Bad {
+                    reason: "no amm pool with base tokens with enough balance".to_string(),
```

Comment:
Nit: not sure about having "with" twice so close to each other here. Could replace the second with "having".